File: integration-tests/src/test/java/dev/capylang/test/ExtensionTest.java
package dev.capylang.test;

import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

class ExtensionTest {
    @Test
    void newPoint3dFromScalars() {
        var result = Extension.newPoint3d(1.0f, 2.0f, 3.0f);

        assertThat(result.x()).isEqualTo(1.0f);
        assertThat(result.y()).isEqualTo(2.0f);
        assertThat(result.z()).isEqualTo(3.0f);
    }

    @Test
    void newPoint3dFromPointSpread() {
        var point = new Extension.Point(1.0f, 2.0f);

        var result = Extension.newPoint3d(point, 3.0f);

        assertThat(result).isEqualTo(new Extension.Point3D(1.0f, 2.0f, 3.0f));
    }

    @Test
    void newPoint3dWithNamedFields() {
        var result = Extension.newPoint3dVer2(4.0f, 5.0f, 6.0f);

        assertThat(result).isEqualTo(new Extension.Point3D(4.0f, 5.0f, 6.0f));
    }

    @Test
    void lengthUsesMostConcreteOverload() {
        assertThat(Extension.length(new Extension.Point(3.0f, 4.0f))).isEqualTo(5.0f);
        assertThat(Extension.length(new Extension.Point3D(1.0f, 2.0f, 2.0f))).isEqualTo(3.0f);
    }

    @Test
    void acceptsPoint3dWherePointIsRequired() {
        assertThat(Extension.onlyPoint2d()).isEqualTo(3.0f);
        assertThat(Extension.testSubstitution(new Extension.Point3D(3.0f, 4.0f, 100.0f))).isEqualTo(5.0f);
    }

    @Test
    void supportsMultipleDataExtensions() {
        var point = new Extension.WeightedLabeledPoint3D(
                1.0f,
                2.0f,
                3.0f,
                "capy",
                4.0f,
                "v1"
        );

        assertThat(Extension.weightedLabeledPointScore(point)).isEqualTo(10.0f);
        assertThat(Extension.weightedLabeledPointLabel(point)).isEqualTo("capy:v1");
        assertThat(Extension.overloadPrefersMostConcrete(point)).isEqualTo((float) Math.sqrt(14.0));
    }

    @Test
    void createsMultiExtendedDataByNamedFields() {
        var result = Extension.newLabeledPoint(1.0f, 2.0f, "origin");

        assertThat(result).isEqualTo(new Extension.LabeledPoint(1.0f, 2.0f, "origin"));
    }

    @Test
    void prefersConcreteOverAnyWhenCoercionsTie() {
        var value = Extension.overloadPrefersConcreteOverAny(new Extension.Point3D(1.0f, 2.0f, 3.0f));

        assertThat(value).isEqualTo("point");
    }
}

File: capy/src/e2e-cfun/js/dev/capylang/test/Extension.test.js
const test = require('node:test');
const { assert, generatedModule } = require('./_support');

test('Extension', () => {
    const extension = generatedModule('dev/capylang/test/Extension.js');
    assert.equal(extension.newPoint3d__float__float__float(1.0, 2.0, 3.0).z, 3.0);
    assert.equal(extension.newPoint3dVer2(4.0, 5.0, 6.0).z, 6.0);
    assert.equal(extension.length2d(new extension.Point({ x: 3.0, y: 4.0 })), 5.0);
    assert.equal(extension.onlyPoint2d(), 3.0);
    assert.equal(extension.testSubstitution(new extension.Point3D({ x: 3.0, y: 4.0, z: 100.0 })), 5.0);
    const weighted = new extension.WeightedLabeledPoint3D({ x: 1.0, y: 2.0, z: 3.0, label: 'capy', weight: 4.0, tag: 'v1' });
    assert.equal(extension.weightedLabeledPointScore(weighted), 10.0);
    assert.equal(extension.weightedLabeledPointLabel(weighted), 'capy:v1');
    assert.equal(extension.overloadAnyPrefersParent(new extension.Point3D({ x: 3.0, y: 4.0, z: 5.0 })), 'point');
    assert.equal(extension.newLabeledPoint(1.0, 2.0, 'origin').label, 'origin');
});
